### PR TITLE
[Kernel] Add MixFP4 W4A16 Marlin inference support

### PR DESCRIPTION
**Pull request by revollllt**
# Add MixFP4 W4A16 Marlin inference support

## Purpose

Add vLLM inference support for MixFP4A16 compressed-tensors checkpoints on the W4A16 Marlin path.

MixFP4 is an extension of NVFP4. For each 16-weight group, MixFP4 compares the reconstruction MSE and chooses whether the elements in that group use the NVFP4 FP4 codebook or the INT4 (E1M2) codebook.

MixFP4 keeps the NVFP4-style 4.5 bit/element storage layout. The packed weights remain 4 bits/element, and each 16-weight group has one FP8 scale byte. MixFP4 reuses the unused sign bit in that FP8 scale byte as the FP4/INT4 selector, so no side metadata is added.

## Implementation

- Add `mixfp4_marlin_gemm` custom op.
- Extend the existing Marlin HMMA path in `csrc/quantization/marlin` rather than adding a separate MixFP4 kernel directory.
- Decode the FP8 scale sign bit as the FP4/INT4 selector in the dequant path.
- Detect canonical format `mixfp4-pack-quantized`; keep legacy `mixed-fp4-int4-pack-quantized` and observer alias compatibility for existing checkpoints.
- Keep ordinary `nvfp4-pack-quantized` routed to the existing NVFP4 path.
- Support fused layers such as QKV and gate-up by keeping per-partition `weight_global_scale` instead of collapsing fused partitions into a single scalar.
- Current tested scope follows the existing Marlin/NVFP4 style path: tensor-parallel size 1, BF16 activation/model dtype, Qwen/Llama-like linear layers, H100 validation.

## Usage

A MixFP4 checkpoint produced by llm-compressor/compressed-tensors can be loaded like other compressed-tensors checkpoints. No separate MixFP4 runtime flag is needed when the checkpoint config contains `quant_method: compressed-tensors` and `format: mixfp4-pack-quantized`.

Python API:

```python
from vllm import LLM, SamplingParams

llm = LLM(model="path/to/mixfp4-checkpoint", dtype="bfloat16")
outputs = llm.generate(
    ["Solve: 17 + 25 ="],
    SamplingParams(max_tokens=32),
)
```

Server CLI:

```bash
vllm serve path/to/mixfp4-checkpoint --dtype bfloat16
```

If a local test checkpoint does not include the usual compressed-tensors quantization metadata, pass the existing compressed-tensors quantization selector explicitly:

```bash
vllm serve path/to/mixfp4-checkpoint --dtype bfloat16 --quantization compressed-tensors
```

## Cross-project workflow

MixFP4 is intended to be used across the compressed-tensors -> llm-compressor -> vLLM stack:

1. `compressed-tensors` defines the canonical `mixfp4-pack-quantized` format, the `MIXFP4A16` preset, and pack/unpack/fake-quant behavior.
2. `llm-compressor` exposes the user-facing quantization recipe: `QuantizationModifier(targets="Linear", scheme="MixFP4A16", ignore=["lm_head"])`.
3. `vLLM` loads the resulting compressed-tensors checkpoint and routes `mixfp4-pack-quantized` W4A16 layers to the MixFP4 Marlin path.

Related PRs:

- compressed-tensors: https://github.com/vllm-project/compressed-tensors/pull/687
- llm-compressor: https://github.com/vllm-project/llm-compressor/pull/2657
- vLLM: https://github.com/vllm-project/vllm/pull/41004

## Shared validation summary

Storage contract:

- Persisted tensors: `weight_packed`, `weight_scale`, `weight_global_scale`.
- `weight_packed` stores 4 bits/element.
- `weight_scale` remains `torch.float8_e4m3fn`, one scale byte per 16 weights; bit 7 stores the FP4/INT4 selector.
- No extra selector tensor or side metadata is emitted.

Accuracy / quality uses lm-eval + vLLM. Math/reasoning uses Qwen math prompts with chat template and math system instruction. MMLU uses lm-eval default `mmlu`, 5-shot, no chat template. WikiText uses lm-eval default `wikitext`; lower PPL is better.

| Model | Quant | GSM8K | MATH500 strict | AIME25 | MMLU | WikiText word PPL |
|---|---|---:|---:|---:|---:|---:|
| Qwen3-8B | BF16 | 93.71% | 76.20% | 70.00% | 74.93% | 13.4838 |
| Qwen3-8B | NVFP4 | 91.58% | 72.00% | 56.67% | 73.55% | 13.7782 |
| Qwen3-8B | MixFP4 | 93.40% | 76.20% | 56.67% | 73.88% | 13.5464 |
| Qwen3-14B | BF16 | 95.07% | 78.80% | 66.67% | 78.82% | 11.6597 |
| Qwen3-14B | NVFP4 | 95.38% | 79.20% | 70.00% | 78.08% | 11.9641 |
| Qwen3-14B | MixFP4 | 94.77% | 76.20% | 66.67% | 78.14% | 11.8100 |

Quality claim: MixFP4 is close to BF16 and improves over NVFP4 on Qwen3-8B for GSM8K, MATH500, MMLU, and WikiText PPL, while tying NVFP4 on AIME25 in the post-fix rerun. On Qwen3-14B, MixFP4 improves WikiText PPL over NVFP4 and is effectively tied on MMLU, while math/reasoning is mixed. AIME25 has only 30 examples, so small-sample deltas should not be overinterpreted.

Latest-vLLM E2E performance on Qwen3-8B with H100 80GB, CUDA 12.8, Torch 2.11, and FA3 enabled:

| Shape | Quant | Wall median ms | Total tok/s median | Decode tok/s median |
|---|---|---:|---:|---:|
| prefill-heavy in=1024 out=32 bs=1 | BF16 | 233.48 | 4522.9 | 148.4 |
| prefill-heavy in=1024 out=32 bs=1 | NVFP4 | 176.25 | 5991.5 | 244.4 |
| prefill-heavy in=1024 out=32 bs=1 | MixFP4 | 187.25 | 5639.6 | 230.1 |
| decode-heavy in=32 out=512 bs=1 | BF16 | 3402.27 | 159.9 | 150.6 |
| decode-heavy in=32 out=512 bs=1 | NVFP4 | 2073.25 | 262.4 | 247.4 |
| decode-heavy in=32 out=512 bs=1 | MixFP4 | 2182.98 | 249.2 | 235.0 |
| batched in=512 out=128 bs=8 | BF16 | 982.90 | 5209.1 | 1141.7 |
| batched in=512 out=128 bs=8 | NVFP4 | 736.97 | 6947.3 | 1837.8 |
| batched in=512 out=128 bs=8 | MixFP4 | 782.96 | 6539.3 | 1751.0 |

MixFP4 is about 5-6% slower than the existing NVFP4 Marlin path in these Qwen3-8B shapes, while remaining substantially faster than BF16 in decode-heavy and batched cases.

## Test Plan

- Run compressed-tensors quantization scheme routing tests for MixFP4 and NVFP4 non-regression.
- Build vLLM with the MixFP4 custom op enabled and verify op registration.
- Run decode-heavy generation smoke for NVFP4 and MixFP4 checkpoints.
- Run E2E benchmark on Qwen3-8B comparing BF16, NVFP4, and MixFP4 on H100.
- Run lm-eval accuracy/PPL validation on Qwen3-8B and Qwen3-14B MixFP4 checkpoints.

## Test Result

- Unit tests: `8 passed, 22 deselected`.
- Build smoke: `mixfp4_marlin_gemm` registered; canonical/legacy MixFP4 formats detected; NVFP4 not mis-routed.
- Decode smoke on Qwen3-8B, input 32, output 512, batch 1:

| Quant | Wall median ms | ITL median ms | Total tok/s median | Decode tok/s median |
|---|---:|---:|---:|---:|
| NVFP4 | 2075.85 | 4.047 | 262.1 | 247.1 |
| MixFP4 | 2193.67 | 4.275 | 248.0 | 233.9 |

- E2E benchmark on Qwen3-8B shows MixFP4 is about 5-6% slower than the existing NVFP4 Marlin path in the tested shapes, while remaining substantially faster than BF16 in decode-heavy and batched cases.
- Accuracy/PPL validation results are listed above.

## Documentation / disclosure

- The usage section above documents how compressed-tensors/llm-compressor/vLLM users can create and load MixFP4 checkpoints.
- This PR was prepared with AI-assisted coding/review; the author reviewed and validated the changes.

## Future work

- Add broader model/checkpoint coverage after the compressed-tensors and llm-compressor MixFP4 PRs are available upstream.
- Investigate a future WGMMA/Triton/CUTLASS path separately from this Marlin integration.

### CODE REVIEW

File: csrc/quantization/marlin/dequant.h [hunk 1]
```diff
@@ -547,6 +547,66 @@ __device__ inline void dequant_fp8_scales<nv_bfloat162, vllm::kFE4M3fn.id()>(
...
+  Out2 += bake_2;
+
+  frag_b[1] = *reinterpret_cast<const nv_bfloat162*>(&Out1);
+  frag_b[0] = *reinterpret_cast<const nv_bfloat162*>(&Out2);
+
+  constexpr uint32_t BITS_2_127 = 254u << 23;
+  const float f_2_127 = *reinterpret_cast<const float*>(&BITS_2_127);
+  const nv_bfloat162 v_2_127 = __float2bfloat162_rn(f_2_127);
+  frag_b[0] = __hmul2(frag_b[0], v_2_127);
+  frag_b[1] = __hmul2(frag_b[1], v_2_127);
```
[the remark this answers is not in the document]
Comment (revollllt) [reply csrc/quantization/marlin/dequant.h:594]: Thanks for the careful check. I added an inline comment explaining the bit-level scale conversion: the MixFP4 path intentionally clears the FP8 sign bit so it can be used as the FP4/INT4 selector, expands the remaining E4M3 scale bits into an un-biased BF16 exponent field, and then multiplies by 2^127 to restore the BF16 exponent bias. The INT4 x4 correction is baked by adding 2 to the BF16 exponent field when the selector bit is set; the remaining FP4 exponent compensation is folded into the Python-side global-scale preprocessing, matching the existing NVFP4 Marlin global-scale handling for BF16.

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w4a16_mixfp4.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        )
+        layer.register_parameter("weight_scale", weight_scale)
+
+    def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+        layer.weight = Parameter(layer.weight_packed.data, requires_grad=False)
+        del layer.weight_packed
+        layer.weight_global_scale = Parameter(
+            1.0 / layer.weight_global_scale.max().to(torch.float32),
+            requires_grad=False,
+        )
```
[the remark this answers is not in the document]
Comment (revollllt) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w4a16_mixfp4.py:104]: For fused layers such as QKV, the checkpoint can carry one `weight_global_scale` per output partition, for example Q, K, and V. Gate-up fused layers are similar, with one scale for gate and one scale for up.

The loading path and the kernel call use different representations:

- During checkpoint loading, vLLM can read per-partition `weight_global_scale` values.
- The Marlin kernel still receives one shared `global_scale`.
- Before calling the kernel, vLLM moves each partition's global-scale ratio into that partition's local scales.

For each weight group, the reconstructed value depends on:

`codebook_value * local_scale * global_scale`

So this adjustment keeps `local_scale * global_scale` unchanged for Q, K, V, and any other fused partition, while avoiding any kernel interface change. The FP4/INT4 choice bit in the local scale byte is preserved during the adjustment, and the unit coverage includes a fused-partition case with different global scales.

**Comment by kylesayrs** [comment]
Looks like an interesting technique that trades decreased performance for higher accuracy. Here's the [paper](https://arxiv.org/abs/2605.31035). @mgoin @DarkLight1337 any thoughts on how difficult this would be to maintain in vLLM? I think this would probably need to be a separate kernel (probably based on humming) rather than a flag on marlin, despite the simplicity.
